### PR TITLE
`cc` Lagrangian and OPDM to Matrix; UHF Bugfix

### PR DESCRIPTION
**Pull request by JonathonMisiewicz**
## Description
`cc` now uses `Matrix` rather than `block_matrix` for the OPDM and Lagrangian. During the course of this transition, I discovered an indexing bug in the conversion of the UHF densities to Matrix. This dates [back to 2016](https://github.com/psi4/psi4/commit/d4bbb98291afe4b500476d0d46c87ded3821bed6). As-is seen from the test values that had to change, this can lead to dramatic errors in properties.

## User API & Changelog headlines

- [x] Fixed a bug where UHF-CCSD properties were incorrect when some orbitals were singly occupied

## Dev notes & details

- [x] `cc` now stores its Lagrangian and OPDM as `Matrix` objects rather than block_matrices rather than write them to disk
- [x] `cc` now uses the wfn based interface, rather than the disk interface, to gradients 
- [x] Adds a finite difference correctness test to  UHF-CCSD one-electron properties
- [x] Much heavier use of slice technology 

## Questions
- [x] Lori, does the test look okay to you?
- [x] Should I do another check on the dipole that changed the most in our test? That change is _massive_.

## Checklist
- [x] `cc` tests pass
- [x] Edited test with incorrect reference values 

## Status
- [x] Ready for review
- [ ] Ready for merge

### CODE REVIEW

File: psi4/src/psi4/cc/ccdensity/sortI_ROHF.cc
Comment (loriab) [inline psi4/src/psi4/cc/ccdensity/sortI_ROHF.cc:90]: valuable explanation, thanks

File: tests/pytests/test_dipoles.py
Comment (loriab) [inline tests/pytests/test_dipoles.py:51]: good test. why not merge it with `test_dipole` since all but the reference repeated?

ah, molecule. can do `mol = h2o_doublet if "reference" not in inp["options"] else h2o_singlet`

File: tests/cc56/input.dat
Comment (loriab) [inline tests/cc56/input.dat:43]: this second one is a jump -- check through cfour sometime?

**Review by loriab — APPROVED** [review]
wow, congratulations on the indexing slog. I can see the value in slices :-)

File: psi4/src/psi4/cc/ccdensity/sortI_RHF.cc
Comment (JonathonMisiewicz) [reply psi4/src/psi4/cc/ccdensity/sortI_RHF.cc:112]: The normalization convention for the Lagrangian that `cc` wrote to **disk** was too large (relative to the **wfn** convention) by a factor of 2. The 2.0 scale on those lines is accounted for by changing the normalization convention from disk to wfn.

File: tests/cc56/input.dat
Comment (JonathonMisiewicz) [reply tests/cc56/input.dat:43]: Sounds good. I don't think Emory has CFOUR, but if we do, I'll be happy to run this before the PR is merged on. I'll check on Monday.

File: tests/pytests/test_dipoles.py
Comment (JonathonMisiewicz) [reply tests/pytests/test_dipoles.py:51]: So, define both `h2o_doublet` and `h2o_singlet` inside the test?

File: psi4/src/psi4/cc/ccdensity/sortI_RHF.cc
Comment (loriab) [reply psi4/src/psi4/cc/ccdensity/sortI_RHF.cc:112]: aha, another factor of 2 that's a fine convention until it collides with another convention :-)

File: tests/pytests/test_dipoles.py
Comment (loriab) [reply tests/pytests/test_dipoles.py:51]: yes, and then pass `properties(..., molecule=mol)`

Comment (loriab) [reply tests/pytests/test_dipoles.py:51]: actually, `mol = h2o_singlet if inp["options"].get("reference", "rhf") == "rhf" else h2o_doublet` would be more robust

**Review by maxscheurer — APPROVED** [review]

**Review by lothian — APPROVED** [review]
It would be good to talk about the reasons for the orbital ordering in ROHF cases (perhaps at PsiCon), but I think the PR looks good.  We have CFOUR access here if that would help with output comparisons.

File: tests/cc56/input.dat
Comment (loriab) [reply tests/cc56/input.dat:43]: I can make cfour generate all three new transition dipoles, so I think this is good to go whenever Jonathon wants to merge.

Orbital ordering whyfors at psicon sounds interesting :-)

**Comment by JonathonMisiewicz** [comment]
I'm happy to talk orbital ordering at PsiCon. One of the points I'll be making in my talk is the importance of keeping this sort of thing documented.

I'll be pushing one more commit momentarily, when build and dipole tests pass again. After that, it'll be good to merge.

**Pull request merged**
